src/i2c/src/i2c.rs

```diff
@@ -0,0 +1,548 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use libc::{c_ulong, ioctl, EADDRINUSE, EADDRNOTAVAIL, EINVAL};
+use std::fs::{File, OpenOptions};
+use std::os::unix::io::AsRawFd;
+use vmm_sys_util::errno::{errno_result, Error, Result};
+
+// The type of the `req` parameter is different for the `musl` library. This will enable
+// successful build for other non-musl libraries.
+#[cfg(target_env = "musl")]
+type IoctlRequest = libc::c_int;
+#[cfg(not(target_env = "musl"))]
+type IoctlRequest = c_ulong;
+
+/// Linux I2C/SMBUS definitions
+/// IOCTL commands
+const I2C_SLAVE: IoctlRequest = 0x0703; // Use this slave address
+const I2C_FUNCS: IoctlRequest = 0x0705; // Get the adapter functionality mask
+const I2C_RDWR: IoctlRequest = 0x0707; // Combined R/W transfer (one STOP only)
+const I2C_SMBUS: IoctlRequest = 0x0720; // SMBus transfer
+
+/// Functions
+const I2C_FUNC_I2C: u64 = 0x00000001;
+const I2C_FUNC_SMBUS_READ_BYTE: u64 = 0x00020000;
+const I2C_FUNC_SMBUS_WRITE_BYTE: u64 = 0x00040000;
+const I2C_FUNC_SMBUS_READ_BYTE_DATA: u64 = 0x00080000;
+const I2C_FUNC_SMBUS_WRITE_BYTE_DATA: u64 = 0x00100000;
+const I2C_FUNC_SMBUS_READ_WORD_DATA: u64 = 0x00200000;
+const I2C_FUNC_SMBUS_WRITE_WORD_DATA: u64 = 0x00400000;
+
+const I2C_FUNC_SMBUS_BYTE: u64 = I2C_FUNC_SMBUS_READ_BYTE | I2C_FUNC_SMBUS_WRITE_BYTE;
+const I2C_FUNC_SMBUS_BYTE_DATA: u64 =
+    I2C_FUNC_SMBUS_READ_BYTE_DATA | I2C_FUNC_SMBUS_WRITE_BYTE_DATA;
+const I2C_FUNC_SMBUS_WORD_DATA: u64 =
+    I2C_FUNC_SMBUS_READ_WORD_DATA | I2C_FUNC_SMBUS_WRITE_WORD_DATA;
+const I2C_FUNC_SMBUS_ALL: u64 =
+    I2C_FUNC_SMBUS_BYTE | I2C_FUNC_SMBUS_BYTE_DATA | I2C_FUNC_SMBUS_WORD_DATA;
+
+/// SMBUS protocol definitions
+/// SMBUS read or write markers
+const I2C_SMBUS_WRITE: u8 = 0;
+const I2C_SMBUS_READ: u8 = 1;
+
+/// SMBus transaction types (size parameter in the above functions)
+const I2C_SMBUS_QUICK: u32 = 0;
+const I2C_SMBUS_BYTE: u32 = 1;
+const I2C_SMBUS_BYTE_DATA: u32 = 2;
+const I2C_SMBUS_WORD_DATA: u32 = 3;
+
+/// As specified in SMBus standard
+const I2C_SMBUS_BLOCK_MAX: usize = 32;
+
+#[repr(C)]
+union I2cSmbusData {
+    byte: u8,
+    word: u16,
+    block: [u8; I2C_SMBUS_BLOCK_MAX + 2],
+}
+
+#[repr(C)]
+struct I2cSmbusIoctlData {
+    read_write: u8,
+    command: u8,
+    size: u32,
+    data: *mut I2cSmbusData,
+}
+
+/// I2C protocol definitions
+pub const I2C_M_RD: u16 = 0x0001; // read data, from slave to master
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+struct I2cMsg {
+    addr: u16,
+    flags: u16,
+    len: u16,
+    buf: *mut u8,
+}
+
+#[repr(C)]
+struct I2cRdwrIoctlData {
+    msgs: *mut I2cMsg,
+    nmsgs: u32,
+}
+
+/// I2C backend definitions
+pub struct I2cReq {
+    pub addr: u16,
+    pub flags: u16,
+    pub len: u16,
+    pub buf: Vec<u8>,
+}
+
+/// I2C adapter and helpers
+#[derive(Debug)]
+struct I2cAdapter {
+    fd: File,
+    bus: u32,
+    smbus: bool,
+}
+
+impl I2cAdapter {
+    fn new(dev: String, dryrun: bool) -> Result<I2cAdapter> {
```

Comment:
Seems "dryrun" is used for tests, but the way is a little inflexible. Should we abstract them into a real device implementation and a mock device implementation? Say:
```
trait I2cAdapterDevice {
}

struct I2cRealDevice {
}

impl I2cAdapterDevice for I2cRealDevice {
}

struct I2cAdapter<B: I2cAdapterDevice> {
}

mod tests {
    struct I2cMockDevice {
    }

    impl I2cAdapterDevice for I2cMockDevice {
    }
}

Reply:
> Seems "dryrun" is used for tests, but the way is a little inflexible. Should we abstract them into a real device implementation and a mock device implementation?

Done